kplutus-pyk/src/kplutus_pyk/genv.py

```diff
@@ -0,0 +1,91 @@
+from typing import Any, List, Tuple
+
+from pyk.kast import (
+    KApply,
+    KAtt,
+    KFlatModule,
+    KImport,
+    KInner,
+    KLabel,
+    KProduction,
+    KRewrite,
+    KRule,
+    KTerminal,
+    KToken,
+    KVariable,
+)
+from pyk.ktool import KRun
+
+_lam = '(lam__)_UPLC-SYNTAX_Term_UplcId_Term'
+_app = '[__]_UPLC-SYNTAX_Term_Term_TermList'
+
+
+def extract_genv(pgm: KInner, krun: KRun) -> Tuple[KInner, KInner]:
+    run_node = pgm.to_dict()
+    current = run_node["args"][1]
+    nest: List[Any] = []
+
+    # Recurse down the program and find where the global environment
+    # definition ends, and the contract body begins
+    while True:
+        try:
+            args = current['args']
+            label = current['label']['name']
+        except KeyError:
+            # Break if the current term is a token
+            break
```

Comment:
Consider adding a more direct check (based on the comment, `KeyError` will be raised exactly when `current` is a token).